materialized: support Materialize Cloud authentication

mjibson opened the pull request:
Add a new cli flag to specify a tenant and JWK causing pgwire and the
http endpoint to require a valid JWT from Materialize Cloud. We trust
the validated username, and will create it if it does not exist.

The JWT comes with expiry, refresh, and role information that we currently
ignore, but will use in the future. The cloud sdk has a function that
does some of this, but omits the expiry information, so isn't good enough
for use here.

### Motivation

  * This PR adds a feature that has not yet been specified.

Support password auth backed by the cloud site to allow users to connect without a client cert.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

antifuchs commented on src/materialized/src/bin/materialized/main.rs:
```diff
@@ -475,6 +486,25 @@ fn run(args: Args) -> Result<(), anyhow::Error> {
+            // Attempt to base64 decode the jwk, otherwise keep it as is. This is
+            // unambiguous because the standard base64 alphabet does not contain dash,
+            // which is present in the key.
+            Some(jwk) => base64::decode(&jwk).unwrap_or_else(|_| jwk.as_bytes().to_vec()),
```
Oh, interesting! Cloud currently passes the key as a literal PEM-encoded string on the environment (with newlines and all), no base64.

antifuchs replied:
(Oh, I just realized that this does an `unwrap_or_else`, reading the key verbatim if base64 fails. Looks good.

antifuchs commented on src/pgwire/src/protocol.rs:
```diff
@@ -156,6 +159,35 @@ where
+        conn.flush().await?;
+        match conn.recv().await? {
+            Some(FrontendMessage::Password { password }) => {
+                match auth.validate_password(&password).await {
```
And here's that load-bearing if statement (-:

antifuchs commented on src/materialized/src/bin/materialized/main.rs:
```diff
@@ -475,6 +486,25 @@ fn run(args: Args) -> Result<(), anyhow::Error> {
+        };
+        Some(MzCloudParams {
+            admin_api_token_url: args.mz_cloud_api_token_url.unwrap_or_else(|| {
+                "https://admin.cloud.materialize.com/identity/resources/auth/v1/api-token"
```
Hrmmmmm, I'm not loving the hardcoded URL - this makes it impossible to authenticate users in staging / dev stacks. Could we make this configurable via a commandline parameter?

antifuchs replied:
Oh, I seeeeee. In that case, I think we can drop that URL - it's unlikely to be real anyway (:

mjibson replied:
Drop as in...make this a required instead of optional parameter?

antifuchs replied:
Yeah, sorry for being unclear. I think it doesn't make sense to provide a default URL here - making it mandatory would be the right thing to do, I think.

benesch commented on src/materialized/src/http.rs:
```diff
@@ -305,3 +351,43 @@ impl hyper::service::Service<Request<Body>> for ThirdPartyServer {
+    req: &Request<Body>,
+    mzcloud: &MzCloudAuthentication,
+) -> Result<String, anyhow::Error> {
+    let auth = match req.headers().get(AUTHORIZATION) {
+        Some(auth) => auth,
+        None => anyhow::bail!("no Authorization header"),
+    };
+    // There doesn't seem to be a way to get the Basic or Bearer struct out of
+    // the Authorization without doing a second decode. We can't just call those
+    // decodes first either, because they debug_assert that the HeaderValue starts
+    // with Basic or Bearer. Is there a better way to do this?
+    let (http_user, jwt): (Option<String>, String) =
+        if Authorization::<Basic>::decode(&mut once(auth)).is_ok() {
+            let basic = Basic::decode(auth).unwrap();
+            let jwt = mzcloud
+                .exchange_password_for_token(basic.password())
+                .await?;
+            (Some(basic.username().to_string()), jwt.access_token)
+        } else if Authorization::<Bearer>::decode(&mut once(auth)).is_ok() {
+            let bearer = Bearer::decode(auth).unwrap();
+            (None, bearer.token().to_string())
+        } else {
+            anyhow::bail!("unexpected authorization scheme");
+        };
```
```suggestion
    let (http_user, jwt) = if let Some(basic) = req.headers().typed_get::<Authorization<Basic>>() {
        let jwt = mzcloud
            .exchange_password_for_token(basic.0.password())
            .await?;
        (Some(basic.0.username().to_string()), jwt.access_token)
    } else if let Some(basic) = req.headers().typed_get::<Authorization<Bearer>>() {
        (None, basic.0.token().to_string())
    } else {
        anyhow::bail!("invalid authorization");
    };
```

mjibson replied:
Ahhh HeaderMapExt. Yes thanks. I also didn't realize that one can use `.entry()` which returns this interface instead of having to make a vec yourself. Rust can be obtuse if you don't happen to understand multiple traits at once.

benesch reviewed (commented):
Didn't make it very far yet, but I'm working on it!

antifuchs commented:
Ran this on my cloud test system with no custom parameters (using only the defaults provided by mzcloud), and got this error message:

```console
materialize-65131269621792946 error: The following required arguments were not provided:
materialize-65131269621792946     --mz-cloud-jwk <MZ_CLOUD_JWK>
materialize-65131269621792946     --mz-cloud-api-token-url <MZ_CLOUD_API_TOKEN_URL>
materialize-65131269621792946 USAGE:
materialize-65131269621792946     materialized --tls-key <PATH> --tls-cert <PATH> --mz-cloud-jwk <MZ_CLOUD_JWK> --mz-cloud-api-token-url <MZ_CLOUD_API_TOKEN_URL> --mz-cloud-tenant <MZ_CLOUD_TENANT> --log-fil
materialize-65131269621792946 For more information try --help
```

mjibson commented:
I've added a `--user-authentication` flag which is now the single thing that enables mzcloud auth, the other mzcloud cli options are optional and ignored unless mzcloud auth is specified. From the cloud team, this sounds like it makes their deployments easier because all deployments can always set those 3 env vars, and then optionally set the user auth to mzcloud when needed.

mjibson commented on src/materialized/src/bin/materialized/main.rs:
```diff
@@ -313,6 +313,34 @@ struct Args {
+    )]
+    mz_cloud_api_token_url: Option<String>,
+    /// Specifies the user authentication method, if differing from the default.
+    #[clap(long, hide = true, possible_values = &["mzcloud"])]
```
@antifuchs There's no env var here, forcing this to be a cli arg, but I just realized that maybe your stuff doesn't do custom args easily. Should it also have an env equivalent like the others?

benesch commented:
> I've added a `--user-authentication` flag which is now the single thing that enables mzcloud auth, the other mzcloud cli options are optional and ignored unless mzcloud auth is specified. From the cloud team, this sounds like it makes their deployments easier because all deployments can always set those 3 env vars, and then optionally set the user auth to mzcloud when needed.

Hm, so now there are a bunch of CLI flags that silently do nothing unless you pass the `--user-authentication` flag? I don't like the sound of that. :/ I would prefer that Materialize Cloud learn how to construct valid `materialized` invocations that pass the correct CLI arguments for the desired combination of version + features.

benesch commented on src/mzcloud-auth/src/lib.rs:
```diff
@@ -0,0 +1,115 @@
+}
+
+// TODO: Do we care about the sub? Do we need to validate the sub or other
+// things, even if unused?
```
That reminds me: we should double check in our shared Slack channel with Frontegg if they consider email to be a stable identifier.

mjibson replied:
Can you define stable here? Does it mean that frontegg doesn't allow creating multiple users with the same email? That the email is unchangeable? What about if a user is deleted then recreated with the same email, does that get a new sub (do we care? maybe we don't care about the sub?).

Does it matter if the email address is not whatever stable means? What problems occur if an email changes? mz would get the new email on next login and auto create that user, and any permissions (after we add those) would apply to that address. I suppose one problem could be that if the clientid,secret hasn't changed but the email address did, then there could be two sessions existing at the same time with different usernames but equal passwords, which certainly seems wrong. Ok so maybe the specific thing we would like to know is if a clientid will always return the same email address.

benesch commented on src/materialized/src/lib.rs:
```diff
@@ -174,6 +177,13 @@ pub enum TlsMode {
+    pub admin_api_token_url: String,
+    pub jwk: Vec<u8>,
+    pub tenant_id: String,
+}
```
Hmm, why this separate type? Can't we just embed the `MzCloudAuthentication` in the config directly?

mjibson replied:
Done.

benesch reviewed (commented):
The code here, including the tests, looks great to me. I have a few things we should shake out, though!

 * The code here is using `CLIENT_ID,SECRET` as the password. That means something like `ded5cbdd-0808-4f00-a947-b05d412f8679,f53cdbbc-e383-4760-baae-888c3842fb83`. I think that is too ugly to use as a password—it just looks like a mistake, like someone accidentally misformatted something. My preference is still to use the URL-safe base64-encoded combination (e.g. `iTZ3wqdMTsuATTt_oMYf3GgQTaI0F0QuphTweNrEcL4`), but I know there were some concerns with that (that I don't fully understand). There's a middle ground of just stripping all the extraneous characters: `ded5cbdd08084f00a947b05d412f8679f53cdbbce3834760baae888c3842fb83`
* Can we be less circumspect and just call this "frontegg" everywhere instead of "mzcloud"? "mzcloud" is pretty awkward (is it `mzcloud` or `mz_cloud` or `Mzcloud` or `MzCloud`?) and results in some weird crate names like `mz-mzcloud-auth`. All these problems go away if we just say `frontegg-auth` and `FronteggAuth`. It's not like we're really abstracting over Frontegg here either; if we were to switch from Frontegg to something else, we'd probably just chuck all the code and start over with something that doesn't do API-keys-as-passwords at all.
* Is it correct that both caching JWTs for username/password pairs and invalidating SQL sessions if the JWT fails to refresh are planned as followup work?

antifuchs commented:
> I would prefer that Materialize Cloud learn how to construct valid materialized invocations that pass the correct CLI arguments for the desired combination of version + features.

I think that's what we're working towards: A thing that Matt & I discussed on Friday was that without an explicit opt-in to mzcloud-based auth it's *extremely* easy for users (not even just mzcloud) to opt themselves into a combination of args that that doesn't keep anyone out (e.g. you specify `--tls-mode=require` without a CA arg). The hope with the `--user-authentication` flag is to switch between different modes that lead to correctly error-checked command-line arguments with no way to accidentally leave the door open.

That the mzcloud auth flags don't do anything isn't ideal (and maybe that can be checked by clap too), but I think that's far less terrible than silently accepting accidentally inactive authentication.

benesch commented:
> A thing that Matt & I discussed on Friday was that without an explicit opt-in to mzcloud-based auth it's extremely easy for users (not even just mzcloud) to opt themselves into a combination of args that that doesn't keep anyone out (e.g. you specify --tls-mode=require without a CA arg)

That's just what `--tls-mode=require` does, though? If you want to verify the client cert, you need either `verify-ca` or `verify-full`, and then `--tls-cert` becomes required, as you would want!

ISTM that we should have three CLI flags:

  * --frontegg-tenant
  * --frontegg-api-url
  * --frontegg-jwk

If anyone of those is specified, then the other two should also be required, and the presence of these three arguments implies that Frontegg authentication is active. We can also change the default of `--tls-mode` to _be_ `require` when Frontegg auth is active. Then the common invocations won't mention `--tls-mode` at all and the right thing will happen:

  * `materialized --frontegg-tenant=blah --frontegg-api-url=blah --frontegg-jwk=blah` specifies that you want Frontegg authentication. `--tls-mode=require` is implied, as its the only TLS mode that makes sense.
  * `materialized --tls-ca=ca.crt --tls-cert=materialized.crt --tls-key=materialized.key` specifies that you want TLS client authentication. `--tls-mode=verify-full` is implied, as its the most secure option.

mjibson commented:
Removed `--user-authentication`. Change all mz_cloud named things to be frontegg. Made tls mode default to requires if the frontegg cli flags are specified. Added UUID parsing to tenant parsing so that the capitalization and dashes specified don't matter. Passwords can now be base64 encoded. Change the base64 alphabet for everything here to be the URL safe alphabet.

benesch reviewed (approved):
This is so wildly awesome!

mjibson commented:
Going to wait on cloud changes before merging this.

benesch commented:
Ok in that case I have a few more password-related things to workshop! Comments incoming.

benesch reviewed (commented):
So, here's what I was thinking on the password front. This gives cloud a whole bunch of flexibility in how they want to encode the client ID/secret into a password:

```diff
diff --git a/src/frontegg-auth/src/lib.rs b/src/frontegg-auth/src/lib.rs
index 29fd2faa7..b4fdbe628 100644
--- a/src/frontegg-auth/src/lib.rs
+++ b/src/frontegg-auth/src/lib.rs
@@ -7,16 +7,11 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use base64::{decode_config, URL_SAFE};
+use anyhow::bail;
 use derivative::Derivative;
 use jsonwebtoken::{decode, Algorithm, DecodingKey, Validation};
 use uuid::Uuid;
 
-fn try_base64_decode<T: AsRef<[u8]>>(input: T) -> Vec<u8> {
-    let input = input.as_ref();
-    decode_config(input, URL_SAFE).unwrap_or_else(|_| input.to_vec())
-}
-
 #[derive(Clone, Derivative)]
 #[derivative(Debug)]
 pub struct FronteggAuthentication {
@@ -28,17 +23,12 @@ pub struct FronteggAuthentication {
 
 impl FronteggAuthentication {
     /// Creates a new frontegg auth. `jwk_rsa_pem` is the RSA public key to
-    /// validate the JWTs and can optionally be base64 encoded (with the URL-safe
-    /// alphabet). `tenant_id` must be parseable as a UUID.
+    /// validate the JWTs. `tenant_id` must be parseable as a UUID.
     pub fn new(
         admin_api_token_url: String,
         jwk_rsa_pem: &[u8],
         tenant_id: &str,
     ) -> Result<Self, anyhow::Error> {
-        // Attempt to base64 decode the password, otherwise keep it as is. This is
-        // unambiguous because the URL-safe base64 alphabet does not contain newline,
-        // which is present in the key.
-        let jwk_rsa_pem = try_base64_decode(&jwk_rsa_pem);
         let decoding_key = DecodingKey::from_rsa_pem(&jwk_rsa_pem)?;
         Ok(Self {
             admin_api_token_url,
@@ -47,28 +37,44 @@ impl FronteggAuthentication {
         })
     }
 
-    /// Exchanges a password (clientid and secret) for a jwt token and validates
-    /// it. Password must have the format: "clientid,secret", and can optionally be
-    /// base64 encoded with the URL-safe alphabet.
-    pub async fn validate_password(&self, password: &str) -> Result<Claims, anyhow::Error> {
-        // Attempt to base64 decode the password, otherwise keep it as is. This is
-        // unambiguous because the URL-safe base64 alphabet does not contain comma,
-        // which is present in the password.
-        let password = String::from_utf8(try_base64_decode(password))?;
-        let token = self.exchange_password_for_token(&password).await?;
-        self.validate_access_token(&token.access_token)
-    }
-
-    // Exchanges a password (clientid and secret) for a jwt token. Password must
-    // have the format: "clientid,secret", and can optionally be
-    /// base64 encoded with the URL-safe alphabet.
+    /// Exchanges a password for a JWT token.
+    ///
+    /// Somewhat unusually, the password encodes both the client ID and secret
+    /// for the API key in use. Both the client ID and secret are UUIDs. The
+    /// password can have one of two formats:
+    ///
+    ///   * The URL-safe base64 encoding of the concatenated bytes of the UUIDs.
+    ///
+    ///     This format is a very compact representation (only 43 or 44 bytes)
+    ///     that is safe to use in a connection string without escaping.
+    ///
+    ///   * The concatenated hex-encoding of the UUIDs, with any number of
+    ///     special characters that are ignored.
+    ///
+    ///     This format allows for the UUIDs to be formatted with hyphens, or
+    ///     not, and for the two
     pub async fn exchange_password_for_token(
         &self,
         password: &str,
     ) -> Result<ApiTokenResponse, anyhow::Error> {
-        let (client_id, secret) = match password.split_once(',') {
-            Some(sp) => sp,
-            None => anyhow::bail!("expected comma-delimiter client+secret"),
+        let (client_id, secret) = if password.len() == 43 || password.len() == 44 {
+            // If it's exactly 43 or 44 bytes, assume we have base64-encoded
+            // UUID bytes without or with padding, respectively.
+            let buf = base64::decode_config(password, base64::URL_SAFE)?;
+            let client_id = Uuid::from_slice(&buf[..16])?;
+            let secret = Uuid::from_slice(&buf[16..])?;
+            (client_id, secret)
+        } else if password.len() >= 64 {
+            // If it's more than 64 bytes, assume we have concatenated
+            // hex-encoded UUIDs, possibly with some special characters mixed
+            // in.
+            let mut chars = password.chars().filter(|c| c.is_alphanumeric());
+            let client_id = Uuid::parse_str(&chars.by_ref().take(32).collect::<String>())?;
+            let secret = Uuid::parse_str(&chars.take(32).collect::<String>())?;
+            (client_id, secret)
+        } else {
+            // Otherwise it's definitely not a password format we understand.
+            bail!("invalid password");
         };
         self.exchange_client_secret_for_token(client_id, secret)
             .await
@@ -77,8 +83,8 @@ impl FronteggAuthentication {
     /// Exchanges a client id and secret for a jwt token.
     pub async fn exchange_client_secret_for_token(
         &self,
-        client_id: &str,
-        secret: &str,
+        client_id: Uuid,
+        secret: Uuid,
     ) -> Result<ApiTokenResponse, anyhow::Error> {
         let resp = reqwest::Client::new()
             .post(&self.admin_api_token_url)
@@ -101,7 +107,7 @@ impl FronteggAuthentication {
         )?;
         let msg_tenant = Uuid::parse_str(&msg.claims.tenant_id)?;
         if msg_tenant != self.tenant_id {
-            anyhow::bail!("tenant ids don't match")
+            bail!("tenant ids don't match")
         }
         Ok(msg.claims)
     }
@@ -109,9 +115,9 @@ impl FronteggAuthentication {
 
 #[derive(Debug, serde::Serialize, serde::Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct ApiTokenArgs<'a> {
-    pub client_id: &'a str,
-    pub secret: &'a str,
+pub struct ApiTokenArgs {
+    pub client_id: Uuid,
+    pub secret: Uuid,
 }
 
 #[derive(Debug, serde::Serialize, serde::Deserialize)]
diff --git a/src/materialized/tests/tls.rs b/src/materialized/tests/tls.rs
index bbd3ee907..66f7fd5e2 100644
--- a/src/materialized/tests/tls.rs
+++ b/src/materialized/tests/tls.rs
@@ -402,8 +402,10 @@ fn test_tls() -> Result<(), Box<dyn Error>> {
     let (bad_client_cert, bad_client_key) = bad_ca.request_client_cert("materialize")?;
 
     let tenant_id = Uuid::new_v4().to_string();
+    let client_id = Uuid::new_v4();
+    let secret = Uuid::new_v4();
     let users = HashMap::from([(
-        ("client".to_string(), "secret".to_string()),
+        (client_id.to_string(), secret.to_string()),
         "user@_.com".to_string(),
     )]);
     let encoding_key = EncodingKey::from_rsa_pem(&ca.pkey.private_key_to_pem_pkcs8().unwrap())?;
@@ -438,8 +440,7 @@ fn test_tls() -> Result<(), Box<dyn Error>> {
         &tenant_id,
     )?;
     let frontegg_user = "user@_.com";
-    let frontegg_password = "client,secret";
-    let frontegg_password_base64 = base64::encode_config(frontegg_password, base64::URL_SAFE);
+    let frontegg_password = &format!("{client_id}{secret}");
     let frontegg_basic = Authorization::basic(frontegg_user, frontegg_password);
     let frontegg_header_basic = make_header(frontegg_basic);
 
@@ -540,10 +541,41 @@ fn test_tls() -> Result<(), Box<dyn Error>> {
                 configure: Box::new(|b| Ok(b.set_verify(SslVerifyMode::NONE))),
                 assert: Assert::Success,
             },
-            // Password can be base64 encoded.
+            // Password can be base64 encoded UUID bytes.
             TestCase::Pgwire {
                 user: frontegg_user,
-                password: Some(&frontegg_password_base64),
+                password: {
+                    let mut buf = vec![];
+                    buf.extend(client_id.as_bytes());
+                    buf.extend(secret.as_bytes());
+                    Some(&base64::encode_config(buf, base64::URL_SAFE))
+                },
+                ssl_mode: SslMode::Require,
+                configure: Box::new(|b| Ok(b.set_verify(SslVerifyMode::NONE))),
+                assert: Assert::Success,
+            },
+            // Password can be base64 encoded UUID bytes without padding.
+            TestCase::Pgwire {
+                user: frontegg_user,
+                password: {
+                    let mut buf = vec![];
+                    buf.extend(client_id.as_bytes());
+                    buf.extend(secret.as_bytes());
+                    Some(&base64::encode_config(buf, base64::URL_SAFE_NO_PAD))
+                },
+                ssl_mode: SslMode::Require,
+                configure: Box::new(|b| Ok(b.set_verify(SslVerifyMode::NONE))),
+                assert: Assert::Success,
+            },
+            // Password can include arbitrary special characters.
+            TestCase::Pgwire {
+                user: frontegg_user,
+                password: {
+                    let mut password = frontegg_password.clone();
+                    password.insert(3, '-');
+                    password.insert_str(12, "@#!");
+                    Some(&password.clone())
+                },
                 ssl_mode: SslMode::Require,
                 configure: Box::new(|b| Ok(b.set_verify(SslVerifyMode::NONE))),
                 assert: Assert::Success,
diff --git a/src/pgwire/src/protocol.rs b/src/pgwire/src/protocol.rs
index 3d91c5ff9..4a5e56626 100644
--- a/src/pgwire/src/protocol.rs
+++ b/src/pgwire/src/protocol.rs
@@ -165,7 +165,10 @@ where
         conn.flush().await?;
         match conn.recv().await? {
             Some(FrontendMessage::Password { password }) => {
-                match frontegg.validate_password(&password).await {
+                let res = frontegg.exchange_password_for_token(&password)
+                    .await
+                    .and_then(|res| frontegg.validate_access_token(&res.access_token));
+                match res {
                     Ok(claims) if claims.email == user => {}
                     _ => {
                         return conn
```

benesch commented:
Ah, and another nice simplification I discovered: you can push the UUID parsing into clap and the claims decoding:

```diff
diff --git a/src/frontegg-auth/src/lib.rs b/src/frontegg-auth/src/lib.rs
index 29fd2faa7..4f634113f 100644
--- a/src/frontegg-auth/src/lib.rs
+++ b/src/frontegg-auth/src/lib.rs
@@ -33,7 +33,7 @@ impl FronteggAuthentication {
     pub fn new(
         admin_api_token_url: String,
         jwk_rsa_pem: &[u8],
-        tenant_id: &str,
+        tenant_id: Uuid,
     ) -> Result<Self, anyhow::Error> {
         // Attempt to base64 decode the password, otherwise keep it as is. This is
         // unambiguous because the URL-safe base64 alphabet does not contain newline,
@@ -43,7 +43,7 @@ impl FronteggAuthentication {
         Ok(Self {
             admin_api_token_url,
             decoding_key,
-            tenant_id: Uuid::parse_str(tenant_id)?,
+            tenant_id,
         })
     }
 
@@ -99,8 +99,7 @@ impl FronteggAuthentication {
             &self.decoding_key,
             &Validation::new(Algorithm::RS256),
         )?;
-        let msg_tenant = Uuid::parse_str(&msg.claims.tenant_id)?;
-        if msg_tenant != self.tenant_id {
+        if msg.claims.tenant_id != self.tenant_id {
             anyhow::bail!("tenant ids don't match")
         }
         Ok(msg.claims)
@@ -130,7 +129,7 @@ pub struct ApiTokenResponse {
 pub struct Claims {
     pub exp: i64,
     pub email: String,
-    pub tenant_id: String,
+    pub tenant_id: Uuid,
     pub roles: Vec<String>,
     pub permissions: Vec<String>,
 }
diff --git a/src/materialized/src/bin/materialized/main.rs b/src/materialized/src/bin/materialized/main.rs
index 27c440d20..fb31389d0 100644
--- a/src/materialized/src/bin/materialized/main.rs
+++ b/src/materialized/src/bin/materialized/main.rs
@@ -50,6 +50,7 @@ use mz_ore::metric;
 use mz_ore::metrics::ThirdPartyMetric;
 use mz_ore::metrics::{raw::IntCounterVec, MetricsRegistry};
 use sysinfo::{ProcessorExt, SystemExt};
+use uuid::Uuid;
 
 use self::tracing::MetricsRecorderLayer;
 use materialized::TlsMode;
@@ -324,7 +325,7 @@ struct Args {
         requires_all = &["frontegg-jwk", "frontegg-api-token-url"],
         hide = true
     )]
-    frontegg_tenant: Option<String>,
+    frontegg_tenant: Option<Uuid>,
     /// JWK used to validate JWTs during user authentication as a PEM public
     /// key. Can optionally be base64 encoded with the URL-safe alphabet.
     #[clap(long, env = "FRONTEGG_JWK", requires = "frontegg-tenant", hide = true)]
@@ -506,7 +507,7 @@ fn run(args: Args) -> Result<(), anyhow::Error> {
             FronteggAuthentication::new(
                 args.frontegg_api_token_url.unwrap(),
                 args.frontegg_jwk.unwrap().as_bytes(),
-                &tenant_id,
+                tenant_id,
             )
         })
         .transpose()?;
diff --git a/src/materialized/tests/tls.rs b/src/materialized/tests/tls.rs
index bbd3ee907..8567ca7e7 100644
--- a/src/materialized/tests/tls.rs
+++ b/src/materialized/tests/tls.rs
@@ -306,18 +306,18 @@ fn run_tests<'a>(header: &str, server: &util::Server, tests: &[TestCase<'a>]) {
 // Users is a mapping from (client, secret) -> email address.
 fn start_mzcloud(
     encoding_key: EncodingKey,
-    tenant_id: &str,
+    tenant_id: Uuid,
     users: HashMap<(String, String), String>,
 ) -> Result<MzCloudServer, anyhow::Error> {
     #[derive(Clone)]
     struct Context {
         encoding_key: EncodingKey,
-        tenant_id: String,
+        tenant_id: Uuid,
         users: HashMap<(String, String), String>,
     }
     let context = Context {
         encoding_key,
-        tenant_id: tenant_id.to_string(),
+        tenant_id,
         users,
     };
     async fn handle(context: Context, req: Request<Body>) -> Result<Response<Body>, Infallible> {
@@ -340,7 +340,7 @@ fn start_mzcloud(
             &Claims {
                 exp: i64::MAX,
                 email: email.to_string(),
-                tenant_id: context.tenant_id.to_string(),
+                tenant_id: context.tenant_id,
                 roles: Vec::new(),
                 permissions: Vec::new(),
             },
@@ -401,7 +401,7 @@ fn test_tls() -> Result<(), Box<dyn Error>> {
     let bad_ca = Ca::new()?;
     let (bad_client_cert, bad_client_key) = bad_ca.request_client_cert("materialize")?;
 
-    let tenant_id = Uuid::new_v4().to_string();
+    let tenant_id = Uuid::new_v4();
     let users = HashMap::from([(
         ("client".to_string(), "secret".to_string()),
         "user@_.com".to_string(),
@@ -412,7 +412,7 @@ fn test_tls() -> Result<(), Box<dyn Error>> {
         &Claims {
             exp: i64::MAX,
             email: "user@_.com".to_string(),
-            tenant_id: tenant_id.to_string(),
+            tenant_id,
             roles: Vec::new(),
             permissions: Vec::new(),
         },
@@ -424,18 +424,18 @@ fn test_tls() -> Result<(), Box<dyn Error>> {
         &Claims {
             exp: i64::MAX,
             email: "user@_.com".to_string(),
-            tenant_id: "bad-tenant".to_string(),
+            tenant_id: Uuid::new_v4(),
             roles: Vec::new(),
             permissions: Vec::new(),
         },
         &encoding_key,
     )
     .unwrap();
-    let frontegg_server = start_mzcloud(encoding_key, &tenant_id, users)?;
+    let frontegg_server = start_mzcloud(encoding_key, tenant_id, users)?;
     let frontegg_auth = FronteggAuthentication::new(
         frontegg_server.url,
         &ca.pkey.public_key_to_pem()?,
-        &tenant_id,
+        tenant_id,
     )?;
     let frontegg_user = "user@_.com";
     let frontegg_password = "client,secret";
```

mjibson commented on src/frontegg-auth/src/lib.rs:
```diff
@@ -0,0 +1,136 @@
+        // Attempt to base64 decode the password, otherwise keep it as is. This is
+        // unambiguous because the URL-safe base64 alphabet does not contain newline,
+        // which is present in the key.
+        let jwk_rsa_pem = try_base64_decode(&jwk_rsa_pem);
```
If you are trying to pass this as a cli flag you need a way to remove the dashes and newlines, and base64 is much easier than...whatever you have to do in bash to do it correctly (I never figured it out). Couldn't use env vars at the time because I was deploying this over mzcloud with its extra args thing. We can possibly just remove the base64 thing from this because cloud is going to do it all in env vars and we won't be doing adhoc deployments for testing.

antifuchs replied:
I've got this wired up in cloud to pass the PEM as an environment variable verbatim, it is fairly easy - don't think we need many provisions for cmdline args for our specific use case.

antifuchs commented on src/materialized/src/bin/materialized/main.rs:
```diff
@@ -313,6 +317,26 @@ struct Args {
+    /// The full URL (including path) to the api-token endpoint.
+    #[clap(
+        long,
+        env = "FRONTEGG_API_TOKEN_URL",
```
since the env variables are named `FRONTEGG_TENANT`, `FRONTEGG_JWK` and `FRONTEGG_API_TOKEN_URL`, this won't interfere / attempt to get confused by the env variables we set in cloud currently (we set `MZCLOUD_JWK` and `MZCLOUD_TENANT`); this means we're pretty much fine to merge this.
